File: bench/ruby/pingpong.rb
# Two threads bounce a token N round trips via queues. Checksum = N.
n = (ENV["BENCH_N"] || "100000").to_i
q_to = Queue.new
q_from = Queue.new
t = Thread.new do
  loop do
    m = q_to.pop
    break if m < 0
    q_from.push(m)
  end
end
k = 0
while k < n
  q_to.push(k)
  q_from.pop
  k += 1
end
q_to.push(-1)
t.join
puts k

File: bench/python/pingpong.py
# Two threads bounce a token N round trips via queues. Checksum = N.
import os, threading, queue
N = int(os.environ.get("BENCH_N", "100000"))
q_to = queue.Queue()
q_from = queue.Queue()


def responder():
    while True:
        m = q_to.get()
        if m < 0:
            break
        q_from.put(m)


t = threading.Thread(target=responder)
t.start()
k = 0
while k < N:
    q_to.put(k)
    q_from.get()
    k += 1
q_to.put(-1)
t.join()
print(k)
